### PR TITLE
CVPN-1506 client first packet event

### PR DESCRIPTION
**Pull request by kp-max-li**
## Description
For each client connection, emits a first packet event when the client got the first outside packet.

## Motivation and Context
Indicate that the first packet is received from the server.

## How Has This Been Tested?
Put some `assert!` in the existing integration tests in `lightway-core` to make sure that the event is only emitted once.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] The correct base branch is being used, if not `main`

### CODE REVIEW

File: lightway-client/src/lib.rs [hunk 1]
```diff
@@ -218,6 +218,10 @@ async fn handle_events<A: 'static + Send + EventCallback>(
             | Event::TlsKeysUpdateCompleted => {
                 unreachable!("server only event received");
             }
+
+            Event::FirstPacketReceived => {
+                // Do nothing
```
Comment (xv-thomas-leong) [inline lightway-client/src/lib.rs]: Can you add an `info!` message indicating the event is received?

File: lightway-core/src/connection.rs [hunk 2]
```diff
@@ -678,6 +682,11 @@ impl<AppState: Send> Connection<AppState> {
     /// In case of UDP, it is almost always one frame per packet. With duplicated
     /// UDP packets, count can be 0.
     pub fn outside_data_received(&mut self, pkt: OutsidePacket) -> ConnectionResult<usize> {
+        if matches!(self.mode, ConnectionMode::Client { .. }) && !self.is_first_packet_received {
```
Comment (xv-thomas-leong) [inline lightway-core/src/connection.rs]: Please put the `is_first_packet_received` boolean check in the front

As Rust performs short circuit on boolean operators, by putting the easier check in the front should speed up execution. I'm assuming the boolean check is less expensive than the `matches!` marco.

Ref: https://doc.rust-lang.org/reference/expressions/operator-expr.html#lazy-boolean-operators

Comment (kp-max-li) [reply lightway-core/src/connection.rs]: Good catch. `!self.is_first_packet_received` will be false the majority of the time for the client, so the change will give less overhead.
I will change it. Thanks.

File: lightway-client/src/lib.rs (hunk 1 above)
Comment (kp-max-li) [reply lightway-client/src/lib.rs]: Sure

**Pull request merged**